src/main/scala/com/microsoft/hyperspace/index/CoveringIndex.scala

```diff
@@ -0,0 +1,420 @@
+        // 1. DATA_FILE_ID_COLUMN column contains source data file id for each index record.
+        // 2. If source data is partitioned, all partitioning key(s) are added to index schema
+        //    as columns if they are not already part of the schema.
+        val partitionColumnNames = relation.partitionSchema.map(_.name)
+        val resolvedColumnNames = (resolvedIndexedColumns ++ resolvedIncludedColumns).map(_.name)
+        val missingPartitionColumns =
+          partitionColumnNames
+            .filter(ResolverUtils.resolve(spark, _, resolvedColumnNames).isEmpty)
+            .map(col)
+
+        // File id value in DATA_FILE_ID_COLUMN column (lineage column) is stored as a
+        // Long data type value. Each source data file has a unique file id, assigned by
+        // Hyperspace. We populate lineage column by joining these file ids with index records.
+        // The normalized path of source data file for each record is the join key.
+        // We normalize paths by removing extra preceding `/` characters in them,
+        // similar to the way they are stored in Content in an IndexLogEntry instance.
+        // Path normalization example:
+        // - Original raw path (output of input_file_name() udf, before normalization):
+        //    + file:///C:/hyperspace/src/test/part-00003.snappy.parquet
+        // - Normalized path (used in join):
+        //    + file:/C:/hyperspace/src/test/part-00003.snappy.parquet
+        import spark.implicits._
+        val dataPathColumn = "_data_path"
+        val lineagePairs = relation.lineagePairs(ctx.fileIdTracker)
+        val lineageDF = lineagePairs.toDF(dataPathColumn, IndexConstants.DATA_FILE_NAME_ID)
+
+        sourceData
+          .withColumn(dataPathColumn, input_file_name())
+          .join(lineageDF.hint("broadcast"), dataPathColumn)
+          .select(projectColumns ++ missingPartitionColumns :+ col(
+            IndexConstants.DATA_FILE_NAME_ID): _*)
+      } else {
+        sourceData.select(projectColumns: _*)
+      }
+
+    (indexData, copy(schema = indexData.schema))
+  }
+
+  override def writeIndexData(ctx: IndexerContext, indexDataFrame: DataFrame): Unit = {
+
+    // Run job
+    val repartitionedIndexData = {
+      // We are repartitioning with normalized columns (e.g., flattened nested column).
+      indexDataFrame.repartition(numBuckets, indexedColumns.map(name => col(s"`$name`")): _*)
+    }
+
+    // Save the index with the number of buckets specified.
+    repartitionedIndexData.write
+      .saveWithBuckets(
+        repartitionedIndexData,
+        ctx.indexDataPath.toString,
+        numBuckets,
+        indexedColumns,
+        SaveMode.Overwrite)
+  }
+
+  override def equals(o: Any): Boolean = o match {
+    case that: CoveringIndexMetadata =>
+      indexedColumns == that.indexedColumns &&
+        includedColumns == that.includedColumns &&
+        numBuckets == that.numBuckets
+    case _ => false
```

Comment:
Is it safe to compare only column names & number of buckets?

Reply:
I don't know, just retaining the old logic. If there's a bug, a separate PR should fix it.